crates/ruff/src/cst/matchers.rs

```diff
@@ -205,3 +198,36 @@ pub(crate) fn match_if<'a, 'b>(statement: &'a mut Statement<'b>) -> Result<&'a m
         bail!("Expected Statement::Compound")
     }
 }
+
+/// Given the source code for an expression, return the parsed [`Expression`].
+///
+/// If the expression is not guaranteed to be valid as a standalone expression (e.g., if it may
+/// span multiple lines and/or require parentheses), use [`transform_expression`] instead.
+pub(crate) fn match_expression(expression_text: &str) -> Result<Expression> {
+    match libcst_native::parse_expression(expression_text) {
+        Ok(expression) => Ok(expression),
+        Err(_) => bail!("Failed to extract expression from source"),
+    }
+}
+
+/// Run a transformation function over an expression.
+///
+/// Passing an expression to [`match_expression`] directly can lead to parse errors if the
+/// expression is not a valid standalone expression (e.g., it was parenthesized in the original
+/// source). This method instead wraps the expression in "fake" parentheses, runs the
+/// transformation, then removes the "fake" parentheses.
+pub(crate) fn transform_expression(
+    source_code: &str,
+    func: impl FnOnce(String) -> Result<String>,
+) -> Result<String> {
+    // Wrap the expression in parentheses.
+    let source_code = format!("({source_code})");
+
+    // Run the function on the expression.
+    let mut transformed = func(source_code)?;
```

Comment:
I would expect this method to orchestrate the whole transformation: Meaning it 

* should call `match_expression` on the source and pass the expression to `func`. Ideally, it would pass the unparenthesized expression, but I think this isn't possible because libCST stores the parenthesized as a field rather than having a separate `ParenthesizedExpression` node.
* `func` mutates the expression and returns a new expression.
* it calls `codegen`, returning the unparenthesized expression

This may have the added benefit that it can remove the parentheses by mutating the CST rather than manipulating the returned string.

Reply:
I structured it this way initially but I lost like an hour or two fighting lifetimes. Let me revisit...

Reply:
> This may have the added benefit that it can remove the parentheses by mutating the CST rather than manipulating the returned string.

Unfortunately LibCST doesn't let you do this, AFAICT :(

Reply:
I can't for the life of me figure out how to make the lifetimes work. The issue is in `format_literals.rs` where we already have to use an Arena to solve weird lifetime problems. Now changing the signature of `transform_expression` gives me:

```rust
error[E0597]: `arena` does not live long enough
   --> crates/ruff/src/rules/pyupgrade/rules/format_literals.rs:220:44
    |
210 |     let output = transform_ex(source_code, stylist, |mut expression| {
    |                                                      -------------- has type `libcst_native::Expression<'1>`
...
219 |         let arena = typed_arena::Arena::new();
    |             ----- binding `arena` declared here
220 |         remove_specifiers(&mut item.value, &arena);
    |         -----------------------------------^^^^^^-
    |         |                                  |
    |         |                                  borrowed value does not live long enough
    |         argument requires that `arena` is borrowed for `'1`
...
223 |     })?;
    |     - `arena` dropped here while still borrowed
```

I've tried moving the arena out of the closure, messing with lifetimes, etc.

Reply:
I was able to change this everywhere except `format_literals.rs`.